text/0000-unsized-rvalues.md

```diff
@@ -0,0 +1,203 @@
+- Feature Name: unsized_locals
+- Start Date: 2017-02-11
+- RFC PR: (leave this empty)
+- Rust Issue: (leave this empty)
+
+# Summary
+[summary]: #summary
+
+Allow for local variables, function arguments, and some expressions to have an unsized type, and implement it by storing the temporaries in variably-sized allocas.
+
+Have repeat expressions with a length that captures local variables be such an expression, returning an `[T]` slice.
+
+Provide some optimization guarantees that unnecessary temporaries will not create unnecessary allocas.
+
+# Motivation
+[motivation]: #motivation
+
+There are 2 motivations for this RFC:
+
+1. Passing unsized values, such as trait objects, to functions by value is often desired. Currently, this must be done through a `Box<T>` with an unnecessary allocation.
+
+  One particularly common example is passing closures that consume their environment without using monomorphization. One would like for this code to work:
+
+  ```Rust
+  fn takes_closure(f: FnOnce()) { f(); }
```

Comment:
> but by the same reasoning it'd be possible for the compiler to monomorphize `fn takes_closure(f: Box<FnOnce()>)` and `fn takes_closure(f: &FnOnce())`, neither of which would be expected to produce monomorphized code. Correct me if I'm wrong, but Rust tries to be explicit about monomorphization vs. trait objects, and adding an optimization such as these seems like it would be taking control away from the user.

Yes, it's a tradeoff. Right now Rust chooses to conflate language features with implementation strategy, largely due to intuitions carried over from C++. But there's no inherent reason it *has* to do this. It's a choice. If we find ourselves asking why the easier way is the slower one and why you have to go out of your way and use a more cumbersome syntax to make it not be slow, and then wanting to complicate the language with even more syntax as a consequence, then it might be worth revisiting the choice.

FWIW I think this sort of thing should('ve) been controlled with attributes. It's as if normal function calls were always inlined and method calls were never inlined and now we were wringing our hands about what to do about it, and wanting to make method syntax less convenient so people are more incentivized to use normal functions, instead of letting the compiler do as it sees fit while providing `#[inline]` as a manual override.